### ISSUE TITLE
TaintedFile false positive on `File::files()` since 4.15.4 facade static sinks

### ISSUE THREAD

**Issue by hoetaek** [psalm/laravel-psalm-plugin#1395]
### Description

Since 4.15.4 ("Report taint sinks on facade static calls", #1318), `File::files($directory)` is reported as `TaintedFile` even when `$directory` can never carry request input.

### Minimal shape

```php
class ListChangelogEntriesAction
{
    use \Lorisleiva\Actions\Concerns\AsAction;

    public function all(?string $directory = null): Collection
    {
        $directory ??= config('changelog.path', resource_path('changelogs'));

        return collect(File::files($directory)) // TaintedFile (psalm.dev/255)
            ->filter(fn ($file) => $file->getExtension() === 'md');
    }
}
```

Call sites in the app pass either `null` or a config path; the only request-derived argument anywhere in the chain is an `int $page` obtained via `$request->integer('page', 1)` and passed as a *named* argument to a different parameter of `::run()`.

### Expected

No taint finding: `config()` / `resource_path()` are not taint sources, and no request data reaches `$directory`.

### Actual

`TaintedFile` on the `File::files()` call. 4.15.3 is clean; bisected by downgrading only this plugin.

### Question

What marks `$directory` as a source here? If `handle()`/`run()` parameters of invokable action classes are treated as request-bound sources, named-argument calls (`run(page: …)`) may be mapping taint onto the wrong parameter.

### Versions

- psalm/plugin-laravel: 4.15.6 (repros), 4.15.3 (clean)
- vimeo/psalm: 7.0.0-beta19
- PHP 8.4, Laravel 13

**Comment by alies-dev** [comment]
Thanks for the precise report — the bisect down to 4.15.4 and the observation about the named argument were both what cracked this.

**Your hypothesis was right, and it is an upstream Psalm bug, not a plugin one.** Filed as vimeo/psalm#11923.

Psalm attributes a named argument's taint to the parameter at its **positional index** rather than to the parameter it names. In `ArgumentsAnalyzer::checkArgumentsMatch()` the argument resolves to the correct parameter but is stored under the positional key, and that offset reaches `DataFlowNode::getForMethodArgument()`. The resulting node mixes the two: its **id** uses the written offset while its reported **location** uses the matched parameter. Sinks are keyed by declared parameter index, so the taint lands on the wrong sink.

It is two-sided, and the second half is worth knowing about:

```php
function sinkNamed(string $path = 'safe', string $label = 'x'): void {
    echo $label;                   // html sink
    echo file_get_contents($path); // file sink
}
sinkNamed(label: tainted());
```

reports `TaintedFile` on `$path` (your false positive) **and stays silent on `$label`**, which is the parameter that actually received the tainted value. A byte-identical positional call reports correctly. Type checking is unaffected throughout — only the taint graph mis-routes.

**One correction to the report:** `$request->integer('page', 1)` is not a taint source, so it is not what seeded this. Some other `input()` / `__get()` read in that chain is the actual source, fused onto the named argument at the `::run()` call. If you are curious which, `--taint-analysis` on that file alone will print the flow.

**#1318 is not the cause.** `Filesystem::files()` has carried a `file $directory` sink all along, and the false positive reproduces with no facade at all (`(new Filesystem)->files($directory)`). #1318 only made a pre-existing upstream bug reachable in your code.

**Fixes:** #1396 (merged) suppresses the mis-attributed taint and clears your `File::files()` report. #1409 follows up with a false-negative fix that missed that merge. Both are Psalm-7 / `master`; a `3.x` backport is coming, since Psalm 6.16.1 is affected identically.

The workaround has a documented accepted false-negative surface, tracked in #1406, and the whole handler retires once vimeo/psalm#11923 is fixed upstream.

**Comment by alies-dev** [comment]
Thank you @hoetaek , the fix is released within https://github.com/psalm/psalm-plugin-laravel/releases/tag/v4.15.7. Please let me know whether it works for your cases.

**Comment by hoetaek** [comment]
Thanks for the fast turnaround. I upgraded to v4.15.7 and re-ran against our codebase: **the `TaintedFile` is still reported, at the same location.**

`psalm/plugin-laravel` v4.15.7, `vimeo/psalm` 7.0.0-beta19, PHP 8.4, Laravel 13. With `--no-cache`: v4.15.3 clean, v4.15.7 → 1 error.

### The real path has no named argument in it

With the named-argument attribution fixed, the trace now prints a flow neither of us had in view:

```
$request->input(...)                     a controller with nothing to do with changelogs
  → SomeUnrelatedAction::run($input)     an action with no file access at all
    → AsObject::run(mixed ...$arguments) lorisleiva/laravel-actions, shared trait
      → static::make()->handle(...$arguments)
        → FindEntryAction::handle#2  $directory              <- taint crosses classes here
          → ListEntriesAction::all → File::files()           TaintedFile
```

(Class and controller names generalised — the repository is private. The shape is exact; the reduction below is the same flow with nothing removed.)

Your correction was right — `$request->integer('page', 1)` was not the seed. The seed is an `input()` read in a different controller feeding a different action, one that never touches the filesystem. No named argument appears anywhere on this path.

### Minimal reproduction: no Laravel, no plugin

Reproduces on plain Psalm 7.0.0-beta19 with **no plugin loaded**, so this looks like core rather than the plugin.

`psalm.xml`:

```xml
<?xml version="1.0"?>
<psalm errorLevel="4" resolveFromConfigFile="true"
    xmlns="https://getpsalm.org/schema/config">
    <projectFiles><directory name="src" /></projectFiles>
</psalm>
```

`src/repro.php` — reports `TaintedFile`:

```php
<?php
trait AsObject {
    public static function run(?string $a = null): mixed { return (new self())->handle($a); }
}
class ListTemplatesAction {
    use AsObject;
    public function handle(?string $type = null): string { return $type ?? 'd'; }
}
class FindEntryAction {
    use AsObject;
    public function handle(?string $dir = null): string {
        $dir ??= '/safe';
        return (string) file_get_contents($dir);   // TaintedFile
    }
}
echo ListTemplatesAction::run($_GET['type']);      // the only entry point; it calls the OTHER class
```

```
vendor/bin/psalm --taint-analysis --no-cache --threads=1 --scan-threads=1
```

`FindEntryAction` is never called from anywhere. The trace shows the crossing explicitly:

```
$_GET['type']                    -> call to ListTemplatesAction::run
ListTemplatesAction::run#1  $a   -> call to FindEntryAction::handle    <- wrong class
FindEntryAction::handle#1   $dir -> file_get_contents($dir)
```

(`TaintedSSRF` is reported on the same call and the same path; `file_get_contents` carries both sinks.)

The control: split the one trait into two and nothing else changes — clean, 0 errors.

```php
<?php
trait AsObjectOne {
    public static function run(?string $a = null): mixed { return (new self())->handle($a); }
}
trait AsObjectTwo {
    public static function run(?string $a = null): mixed { return (new self())->handle($a); }
}
class ListTemplatesAction {
    use AsObjectOne;
    public function handle(?string $type = null): string { return $type ?? 'd'; }
}
class FindEntryAction {
    use AsObjectTwo;
    public function handle(?string $dir = null): string {
        $dir ??= '/safe';
        return (string) file_get_contents($dir);   // no issue
    }
}
echo ListTemplatesAction::run($_GET['type']);
```

### The determining variable is the shared trait

Removing one variable at a time:

| variant | `run()` in a shared trait | variadic | `static::` LSB | TaintedFile |
|---|:-:|:-:|:-:|:-:|
| A | yes | yes | yes | 1 |
| B | yes | no | yes | 1 |
| C | no (own `run()` per class) | yes | no | 0 |
| D | yes | yes | no | 1 |
| E | yes | no | no | 1 |
| F | no (two separate traits) | no | no | 0 |

E and F are byte-identical apart from F splitting `AsObject` into `AsObjectOne` / `AsObjectTwo`. E reports, F is clean. Neither variadic nor late static binding is required — dropping both (E) still reports.

So the taint node for a trait method's parameter appears to be keyed on the **declaring trait symbol** rather than on the using class. Every class that `use`s the trait collapses onto one node, and taint entering through one class's `run()` reaches every other class's `handle()`.

An independent corroboration fell out of the same two runs, outside taint analysis entirely: the control reports `PossiblyUnusedMethod: Cannot find any calls to method FindEntryAction::run`, and the shared-trait version does not. With one shared trait, `run()` is a single symbol that *is* called, so the two classes are indistinguishable to the unused-method check as well. Same collapse, different subsystem.

### Why this is load-bearing for Laravel

`lorisleiva/laravel-actions` gives every action the same `AsAction` trait, so `AsObject::run()` is a single node for the whole application: any request input reaching any action's `::run()` is treated as reaching every action's parameters. #1318 turning facade calls into sinks is what made it visible for us — so your framing holds, #1318 only exposed a pre-existing core bug. It just seems to be a different pre-existing bug from #11923.

This is adjacent to the "handle late static binding" item in #1406, but variants D and E indicate LSB is not the variable; the trait symbol is.

Happy to file this separately against `vimeo/psalm` since it reproduces with no plugin, or to reopen this issue — whichever you prefer. I can also post the taint graph node ids if useful.

**Comment by alies-dev** [comment]
Reopening.

v4.15.7 does not fix this. I attributed the report to vimeo/psalm#11923, named-argument taint pinned to the positional index, because of the `run(page: …)` detail. That handler only inspects syntactically named arguments, and your path passes positionally, so it records nothing. Different bug.

The one you hit is vimeo/psalm#11913, open since late July. Your reduction is smaller than the one already there, so I posted it with the graph dump.

Correction to your reading: the node is not keyed on the declaring trait symbol, it carries no method symbol at all. `FunctionLikeAnalyzer.php:1129` calls `DataFlowNode::getForAssignment($function_param_id, $function_param->location)` with no specialization key, so the id is the variable name plus the trait body's file offsets. Psalm re-analyses that body once per using class, both passes produce the same id, and the two classes' edges merge onto it. The argument nodes either side stay class-correct.

Your abstract-base variants reproduce the symptom by a different route, conservative fan-out over virtual dispatch. A fix for #11913 will not clear that shape.

It also loses real findings, which is the part that affects you most:

```php
echo Wrong::run(['other' => $_GET['a']]);
$tmp = ['path' => $_GET['b']];
echo Vuln::run($tmp);   // Vuln::handle does file_get_contents($a['path'])
```

Nothing is reported. Split the trait in two and `TaintedFile` appears. `TaintFlowGraph.php:459` marks the shared node visited on first arrival, keyed on id and taint bits only, and drops the second before the array-key match at `:465` can see it would have matched.

No general fix is available to us. Node identity is assigned before any plugin hook runs, and `AddTaints` / `RemoveTaints` fire at value sites with the class provenance already gone. We do mask it for `Dispatchable` by accident, since our stub leaves `dispatch()` empty, but we ship no `AsAction` stub, so your action classes get nothing from that.

Options: `@psalm-taint-specialize` clears the false positive and drops true positives on the same path, so I would not; a targeted `@psalm-suppress TaintedFile` on the call; or baseline it.

Leaving this open and linked to #11913 so it closes on the upstream fix. Thanks for the bisect and for coming back after 4.15.7.

**Comment by hoetaek** [comment]
Thanks for reopening this and for the correction — both the attribution and my reading of the node id were wrong on my side, and the false negative is the part I had not considered at all.

I measured our own exposure and posted the detail on vimeo/psalm#11913. Short version, since it answers the one thing you left open there:

**The `AsAction` baseline does hold.** A single class using `AsAction` reports `TaintedFile` + `TaintedSSRF` normally, so the double variadic hop is not losing the array-key path by itself. Add a second class on the same trait calling under a different key and the finding disappears; duplicate the trait per class and it comes back. Same with the plugin loaded, which lines up with your note that there is no `AsAction` stub to empty out.

I also checked the harness against your `Dispatchable` repro first and it reproduces both of your results, with and without the plugin.

One thing worth flagging for whoever verifies the eventual fix: whether the false negative fires depends on what else is in `projectFiles`. Same code, same config — silent with 844 files in scope, reported with 623. So neither "it does not reproduce on a real app" nor "the small fixture passes" is a safe signal on its own. Numbers are in the other issue.

On our side this is closed by removing the parameter that carried the taint, so nothing is blocked for us. Leaving this to close on the upstream fix as you suggested.
